### PR TITLE
Balance priorities during reconciliation

### PR DESCRIPTION
**Pull request by idegtiarenko**
When reconciling a balance with a lot of shards on undesired nodes there is a possibility of causing node hot spots due to usage of nodeInterleavedShardIterator. This iterator orders shards based on nodes they are located and order nodes based hash map iteration. This means it tends to pick shards returned first by the iterator.

Depends on https://github.com/elastic/elasticsearch/pull/96025

Related to https://github.com/elastic/elasticsearch/issues/91386

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java:230]: nit: no real need for these checks, they're the first thing that `logger.debug()` does anyway

**Review by DaveCTurner — COMMENTED** [review]
I wonder if we should stop with the interleaving altogether and just try all the shards from the least-recently-touched node until we find one to move, then move on to the next-least-recently-touched node and try all _its_ shards, and so on.

**Comment by idegtiarenko** [comment]
>I wonder if we should stop with the interleaving altogether and just try all the shards from the least-recently-touched node until we find one to move, then move on to the next-least-recently-touched node and try all its shards, and so on.

We need to handle the case when it is allowed to move more shards then there are nodes in cluster (or somehow to return to the first node again)

**Comment by DaveCTurner** [comment]
> We need to handle the case when it is allowed to move more shards then there are nodes in cluster (or somehow to return to the first node again)

Yes, we'd keep iterating until we've considered moving every shard on every node.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java:469]: I am planning to move this to a top-level class and add some unit tests for this

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java:502]: nit: IMO we should update `moveOrdering` here rather than relying on the caller calling both `dePrioritizeNode` and `recordAllocation`.

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java:1044]: This test looks quite general but is in fact very specific: it needs the initial allocation of shards to be such that all the shards on each initial node must move to the same target node. It's then saying that we round-robin through the nodes rather than focussing on a single node's shards. I think I'd prefer it to be more explicit about this - rather than saying that the cluster never sees a hotspot, I think we should count the number of times a shard moves off each node and assert that these numbers don't get too far apart.

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/OrderedShardsIteratorTests.java
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/OrderedShardsIteratorTests.java:51]: nit: let us use immutable instead, seems more inline with the read-only nature of the iterator:

```suggestion
        var iterator = OrderedShardsIterator.create(RoutingNodes.immutable(routing, nodes), ordering);
```

Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/OrderedShardsIteratorTests.java:51]: nit: let us use immutable instead, seems more inline with the read-only nature of the iterator:

```suggestion
        var iterator = OrderedShardsIterator.create(RoutingNodes.immutable(routing, nodes), ordering);
```

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/OrderedShardsIteratorTests.java
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/OrderedShardsIteratorTests.java:76]:
```suggestion
        var iterator = OrderedShardsIterator.create(RoutingNodes.immutable(routing, nodes), ordering);
```

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/OrderedShardsIteratorTests.java
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/OrderedShardsIteratorTests.java:44]: Can we add an extra index for node-2 to see that that extra index comes out before the shards on node-1? I.e., that it does not return one shard from each node like the interleaved one.

**Review by henningandersen — COMMENTED** [review]
This looks good to me. I left a number of minor comments.

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/OrderedShardsIteratorTests.java
Comment (idegtiarenko) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/OrderedShardsIteratorTests.java:51]: While iterator is not performing any changes I wanted to test it it with mutable RoutingNodes as this is what we supply it in the real code

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java:1109]: I agree - I think we should remove the fully-reconciled nodes at the top of the loop, rather than doing it after the first assertion.

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java:1049]: Is there any reason not to go to 1 here? I.e.:

```suggestion
        int shardsPerNode = randomIntBetween(1, 15);
```

**Review by henningandersen — APPROVED** [review]
LGTM.

**Review by DaveCTurner — APPROVED** [review]
LGTM2

Did we also agree to backport this to 8.8? I think we should.

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java
Comment (idegtiarenko) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java]: Sounds good, I updated the message to `Reconciling nodes should all have same amount (max 1 delta) of moves`

**Comment by idegtiarenko** [comment]
>Did we also agree to backport this to 8.8? I think we should.

I think it is worth backporting while might look for other allocator improvements.
I will open a backport pr, but will not merge it automatically until we make a final decision.

**Comment by idegtiarenko** [comment]
@elasticsearchmachine please run elasticsearch-ci/bwc

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java]: I think that if we allowed 3 outgoing recoveries, we risked one node having just one shard left to move and another having 3 shards left to move. In that case, we'd have a delta of 2 even when we round-robin. Did I misunderstand something or does the check here not rely on max doing 2 outgoing recoveries at a time?

Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java]: If we allowed 3 outgoing recoveries I think we would risk one node having one outgoing relocation and the other node having 3 outgoing relocations, if these were the very last pending relocations needed. This would happen in one `DesiredBalanceReconciler.run` round, i.e., we would not remove from the `totalOutgoingMoves` before after incrementing the counts and validating - and I think that would make the test fail?

I should probably try this out....

**Pull request merged**
